documentation/wiki/MSBuild-overview.md

```diff
@@ -0,0 +1,60 @@
+- `Project` / `ProjectInstance`: entrypoint APIs for working with MSBuild evaluations.
+- `BuildManager`: entrypoint API for executing targets in MSBuild.
+  - - Project build stack: RequestBuilder -> TargetBuilder -> TaskBuilder. The RequestBuilder is responsible for evaluating and running targets on a project. The TargetBuilder is responsible for running the target execution stack. The TaskBuilder is responsible for running individual tasks.
+- MSBuild nodes
+  - MSBuild distributes work across multiple processes.
+  - All processes live on the same machine.
+  - Types of nodes:
+    - A main node, called the "BuildManager node" or the "Scheduler node". It contains BuildManager instances and its main role is to coordinate the entire build. It's a big state machine that decides when to create other nodes and on which nodes to build projects.
+    - Multiple worker nodes whose single responsibility is to build projects. A worker node builds only one project at a time (99% true).
+  - The process hosting the single BuildManager node can also host one worker node (called the "inproc node"). This is done for perf reasons, to avoid serialization costs. When the BuildManager node decides to spawn a new Worker node, it creates a new process for it. These are called out of process nodes, or "oop nodes". So Worker nodes can either be inproc nodes (a single one), or oop nodes (multiple).
+
+# MSBuild execution modes
+- Building from cmdline (type msbuild.exe)
+    - MSBuild is responsible for discovering project dependencies, and building them in a depth first search manner
+    - There is no static graph of project dependencies. MSBuild does not know ahead of time what work it needs to do. It does a Just In Time DFS descent over the project references.
+    - references form a DAG. A -> {B, C}, B -> {C, D}
+    - there's a convention on how to declare and build references
+      - documented here: https://github.com/dotnet/msbuild/blob/main/documentation/ProjectReference-Protocol.md
+      - mostly implemented here: https://github.com/dotnet/msbuild/blob/main/src/Tasks/Microsoft.Common.CurrentVersion.targets
+- Building from Visual Studio (VS) (pressing ctrl+shift+B, or the context menu `Build -> Build Solution`)
+    - VS implements an object model for an MSBuild dependency graph
+    - VS takes over the responsibility of handling the graph dependencies and constructs its own graph
+    - It bypasses MSBuild's scheduler and "orders" MSBuild to build a single project at a time, ignoring its references
+    - It builds the graph bottom up in reverse toposort order (starting from nodes without references and going up).
+    - Fast up to date checks:  VS hardcodes relevant project file inputs for each language SDK. (for example it knows that C# projects read global.json). It uses these hardcoded rules to skip building projects. (whereas quickbuild does this in a generic non hardcoded way by observing actual IO inputs and outputs)
```

Comment:
It's not that VS has hardcoded logic. It's that VS ran the evaluation step described above at some earlier time, so knows those files advertised by the project as being build inputs and outputs. This allows VS to more quickly scan those timestamps and skip calling MSBuild when no build is needed.

Reply:
That's true, but the VS rules on what inputs to extract from a project evaluation are still hardcoded rules (for example: look at @(Compile), look at @(Content), look at global.json, etc). During actual runtime, msbuild tasks may read file inputs not covered by the VS rules (leading to underbuilds, a correctness issue), or may read less file inputs than the VS rules (leading to overbuilds, wasted time). At best, VS can only play catch-up with what happens at runtime. This is opposed to a monitored build runtime that captures the concrete IO that's performed, thus being accurate by definition (or at least as good as the monitoring APIs).